src/blockrelay/mempool_sync.cpp

```diff
@@ -0,0 +1,404 @@
+        inputData.tx = tx;
+        inputData.nodeId = pfrom->id;
+        EnqueueTxForAdmission(inputData);
+    }
+
+    LOG(MPOOLSYNC, "Recovered %d txs from peer=%s via mempool sync\n", mempoolSyncTx.vTx.size(), pfrom->GetLogName());
+
+    // We should not receive any future messages related to this round of synchronization
+    mempoolSyncRequested[pfrom].completed = true;
+
+    return true;
+}
+
+void GetMempoolTxHashes(std::vector<uint256> &mempoolTxHashes)
+{
+    {
+        READLOCK(orphanpool.cs);
+        for (auto &kv : orphanpool.mapOrphanTransactions)
+        {
+            mempoolTxHashes.push_back(kv.first);
+        }
+    }
+
+    std::vector<uint256> memPoolHashes;
+    mempool.queryHashes(memPoolHashes);
+
+    for (const uint256 &hash : memPoolHashes)
+    {
+        mempoolTxHashes.push_back(hash);
+    }
+}
+
+CMempoolSyncInfo GetMempoolSyncInfo()
+{
+    // We need the number of transactions in the mempool and orphanpools but also the number
+    // in the txCommitQ that have been processed and valid, and which will be in the mempool shortly.
+    uint64_t nCommitQ = 0;
+    {
+        boost::unique_lock<boost::mutex> lock(csCommitQ);
+        nCommitQ = txCommitQ->size();
+    }
+
+    uint64_t nTxInMempool = mempool.size() + orphanpool.GetOrphanPoolSize() + nCommitQ;
+    uint64_t nMempoolMaxTxBytes = GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000;
+    uint64_t nSatoshiPerK = minRelayTxFee.GetFeePerK();
+
+    // Form SipHash keys
+    uint64_t seed = GetRand(std::numeric_limits<uint64_t>::max());
+    CDataStream stream(SER_NETWORK, PROTOCOL_VERSION);
+    stream << seed;
+    CSHA256 hasher;
+    hasher.Write((unsigned char *)&(*stream.begin()), stream.end() - stream.begin());
+    uint256 shorttxidhash;
+    hasher.Finalize(shorttxidhash.begin());
+    uint64_t shorttxidk0 = shorttxidhash.GetUint64(0);
+    uint64_t shorttxidk1 = shorttxidhash.GetUint64(1);
+
+    // Calculate how many bytes of space remain in the mempool
+    uint64_t nRemainingMempoolTxBytes = nMempoolMaxTxBytes;
+    {
+        READLOCK(mempool.cs);
+        for (const CTxMemPoolEntry &e : mempool.mapTx)
+        {
+            nRemainingMempoolTxBytes += e.GetTx().GetTxSize();
+        }
+    }
+
+    return CMempoolSyncInfo(nTxInMempool, nRemainingMempoolTxBytes, shorttxidk0, shorttxidk1, nSatoshiPerK);
+}
+
+uint64_t NegotiateMempoolSyncVersion(CNode *pfrom)
+{
+    uint64_t peerMin = pfrom->xVersion.as_u64c(XVer::BU_MEMPOOL_SYNC_MIN_VERSION_SUPPORTED);
+    uint64_t selfMin = mempoolSyncMinVersionSupported.Value();
+    uint64_t peerMax = pfrom->xVersion.as_u64c(XVer::BU_MEMPOOL_SYNC_MAX_VERSION_SUPPORTED);
+    uint64_t selfMax = mempoolSyncMaxVersionSupported.Value();
+
+    uint64_t upper = (uint64_t)std::min(peerMax, selfMax);
+    uint64_t lower = (uint64_t)std::max(peerMin, selfMin);
+
+    if (lower > upper)
+        throw std::runtime_error("Sender and receiver support incompatible mempool sync versions");
```

Comment:
it might be useful here to instead of throwing an error to send some sort of reject message back to the requesting peer and add them to a "non negotiable" list for the duration of the connection so that we don't turn around and later request a sync from them just to have the same error happen. 

when a node receives the reject message it should add the peer that sent it to its own "non negotiable" list to avoid future requests to sync with that node and immediately try to negotiate a mempool sync with another peer since this sync failed.

Reply:
I'm going to paste here what we agreed upon offline. 

Before sending any mempool sync requests, a peer should: (1) check that the BU_MEMPOOL_SYNC xversion value is set to 1 and (2) check that its mempool sync version is compatible with the version of the would-be responder (again using xversion information).